sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreCheckpointFormatV2Suite.scala

```diff
@@ -0,0 +1,1103 @@
+
+      Seq(0, 1).foreach {
+        partitionId =>
+          val stateStoreCkptIds = pickedCheckpointInfoList
+            .filter(_.partitionId == partitionId).map(_.stateStoreCkptId)
+          val baseStateStoreCkptIds = pickedCheckpointInfoList
+            .filter(_.partitionId == partitionId).map(_.baseStateStoreCkptId)
+
+          // Verify lineage for each partition across batches. Below should satisfy because
+          // these ids are stored in the following manner:
+          // stateStoreCkptIds: id3, id2, id1
+          // baseStateStoreCkptIds:  id2, id1, None
+          // Below checks [id2, id1] are the same,
+          // which is the lineage for this partition across batches
+          assert(stateStoreCkptIds.drop(1).iterator
+            .sameElements(baseStateStoreCkptIds.dropRight(1)))
+      }
+
+      val versionToUniqueIdFromStateStore = Seq(1, 2).map {
+        batchVersion =>
+          val res = pickedCheckpointInfoList
+            .filter(_.batchVersion == batchVersion).map(_.stateStoreCkptId.get)
+
+          // batch Id is batchVersion - 1
+          batchVersion - 1 -> res.toArray
+      }.toMap
+
+      val commitLogPath = new Path(
+        new Path(checkpointDir.getAbsolutePath), "commits").toString
+
+      val commitLog = new CommitLog(spark, commitLogPath)
+      val metadata_ = commitLog.get(Some(0), Some(1)).map(_._2)
+
+      val versionToUniqueIdFromCommitLog = metadata_.zipWithIndex.map { case (metadata, idx) =>
+        // Use stateUniqueIds(0) because there is only one state operator
+        val res2 = metadata.stateUniqueIds(0).map { uniqueIds =>
+          uniqueIds
+        }
+        println("wei== res2")
+        res2.foreach(x => for (elem <- x) {
+          println(elem)
+        })
+        idx -> res2
+      }.toMap
+
+      versionToUniqueIdFromCommitLog.foreach {
+        case (version, uniqueIds) =>
+          versionToUniqueIdFromStateStore(version).sameElements(uniqueIds)
+      }
+    }
+  }
+
+  // This test verifies when there are task retries, the unique ids actually stored in
+  // the commit log is the same as those recorded by CkptIdCollectingStateStoreWrapper
+  testWithCheckpointInfoTracked(s"checkpointFormatVersion2 validate ID - task retry") {
+    withTempDir { checkpointDir =>
+      val inputData = MemoryStream[Int]
+      val aggregated =
+        inputData
+          .toDF()
+          .groupBy($"value")
+          .agg(count("*"))
+
+      val writer = (ds: DataFrame, batchId: Long) => {
+        val _ = ds.rdd.filter { x =>
+          val context = TaskContext.get()
+          // Retry in the first attempt
+          if (context.attemptNumber() == 0) {
+            throw new RuntimeException(s"fail the task at " +
+              s"partition ${context.partitionId()} batch Id: $batchId")
+          }
+          x.length >= 0
+        }.collect()
+      }
+
+      val query = aggregated.writeStream
+        .foreachBatch(writer)
```

Comment:
We are betting foreachbatch will run in the same task as the stateful operator. I assume you tested and it is right, but it is not robust enough to maintain as the underlying implementation can change. I think is is slightly better to do the close in ForEach() rather than foreachBatch().

Reply:
I added another one using foreach

Reply:
Then do we still need this one? Can we delete it?

Reply:
Sure we can if you think it's better? Personally I believe more tests don't hurt but up to you